From bd7bc7ae9f46a8678c05dca78081033d6e2af497 Mon Sep 17 00:00:00 2001
From: CN_SZTL <cnsztl@project-openwrt.eu.org>
Date: Sat, 27 Feb 2021 15:05:31 +0800
Subject: [PATCH] cli: bump to v2.3.0

Signed-off-by: CN_SZTL <cnsztl@project-openwrt.eu.org>
---
 cli/brook/main.go | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cli/brook/main.go b/cli/brook/main.go
index 1f715ea..3c88eb2 100644
--- a/cli/brook/main.go
+++ b/cli/brook/main.go
@@ -28,7 +28,7 @@ import (
 	_ "net/http/pprof"
 
 	"github.com/txthinking/brook"
-	"github.com/urfave/cli"
+	"github.com/urfave/cli/v2"
 )
 
 var debug bool
-- 
2.17.1

