--- a/src/r8126_n.c
+++ b/src/r8126_n.c
@@ -13016,6 +13016,7 @@ rtl8126_init_board(struct pci_dev *pdev,
         void __iomem *ioaddr;
         struct net_device *dev;
         struct rtl8126_private *tp;
+        const char *devname = of_get_property(pdev->dev.of_node, "label", NULL);
         int rc = -ENOMEM, i, pm_cap;
 
         assert(ioaddr_out != NULL);
@@ -13030,6 +13031,9 @@ rtl8126_init_board(struct pci_dev *pdev,
                 goto err_out;
         }
 
+        if (devname)
+                strscpy(dev->name, devname, IFNAMSIZ);
+
         SET_MODULE_OWNER(dev);
         SET_NETDEV_DEV(dev, &pdev->dev);
         tp = netdev_priv(dev);
