From 74265362c4a44e2abe6594f263973634092d5c6d Mon Sep 17 00:00:00 2001
From: CN_SZTL <cnsztl@project-openwrt.eu.org>
Date: Fri, 29 Jan 2021 05:39:16 +0000
Subject: [PATCH] main.cpp: fix typo error of getting password

Signed-off-by: CN_SZTL <cnsztl@project-openwrt.eu.org>
---
 main.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/main.cpp b/main.cpp
index e7035be..47df5d0 100644
--- a/main.cpp
+++ b/main.cpp
@@ -25,7 +25,7 @@ int main(int argc, char** argv) {
 
 	bool auth{ result["1"].count() != 0 };
 	bool user{ result["u"].count() != 0 };
-	bool password{ result["p"].count() != 0 };
+	bool password{ result["P"].count() != 0 };
 
 	if (user ^ password) {
 		fprintf(stderr, "error: user and password must be used together\n");
-- 
2.17.1

