--- a/drivers/net/ethernet/realtek/r8169_main.c
+++ b/drivers/net/ethernet/realtek/r8169_main.c
@@ -5260,6 +5260,7 @@ static int rtl_init_one(struct pci_dev *
 	int jumbo_max, region, rc;
 	enum mac_version chipset;
 	struct net_device *dev;
+	const char *devname = of_get_property(pdev->dev.of_node, "label", NULL);
 	u32 txconfig;
 	u16 xid;
 
@@ -5267,6 +5268,9 @@ static int rtl_init_one(struct pci_dev *
 	if (!dev)
 		return -ENOMEM;
 
+	if (devname)
+		strlcpy(dev->name, devname, IFNAMSIZ);
+
 	SET_NETDEV_DEV(dev, &pdev->dev);
 	dev->netdev_ops = &rtl_netdev_ops;
 	tp = netdev_priv(dev);
