--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -278,6 +278,11 @@ config INTEL_XWAY_PHY
 	  PEF 7061, PEF 7071 and PEF 7072 or integrated into the Intel
 	  SoCs xRX200, xRX300, xRX330, xRX350 and xRX550.
 
+config JLSEMI_PHY
+	tristate "JLSemi PHYs"
+	help
+	  Currently supports the JLSemi PHYs.
+
 config LSI_ET1011C_PHY
 	tristate "LSI ET1011C PHY"
 	help
--- a/drivers/net/phy/Makefile
+++ b/drivers/net/phy/Makefile
@@ -75,6 +75,8 @@ obj-$(CONFIG_DP83TD510_PHY)	+= dp83td510
 obj-$(CONFIG_FIXED_PHY)		+= fixed_phy.o
 obj-$(CONFIG_ICPLUS_PHY)	+= icplus.o
 obj-$(CONFIG_INTEL_XWAY_PHY)	+= intel-xway.o
+obj-$(CONFIG_JLSEMI_PHY) += jlsemiphy.o
+jlsemiphy-objs := jlsemi.o jlsemi-core.o
 obj-$(CONFIG_LSI_ET1011C_PHY)	+= et1011c.o
 obj-$(CONFIG_LXT_PHY)		+= lxt.o
 obj-$(CONFIG_MARVELL_10G_PHY)	+= marvell10g.o
