--- a/drivers/net/ethernet/realtek/r8169_main.c
+++ b/drivers/net/ethernet/realtek/r8169_main.c
@@ -21,6 +21,7 @@
 #include <linux/in.h>
 #include <linux/io.h>
 #include <linux/ip.h>
+#include <linux/of.h>
 #include <linux/tcp.h>
 #include <linux/interrupt.h>
 #include <linux/dma-mapping.h>
@@ -5315,12 +5316,16 @@ static int rtl_init_one(struct pci_dev *
 	int jumbo_max, region, rc;
 	enum mac_version chipset;
 	struct net_device *dev;
+	const char *devname = of_get_property(pdev->dev.of_node, "label", NULL);
 	u16 xid;
 
 	dev = devm_alloc_etherdev(&pdev->dev, sizeof (*tp));
 	if (!dev)
 		return -ENOMEM;
 
+	if (devname)
+		strlcpy(dev->name, devname, IFNAMSIZ);
+
 	SET_NETDEV_DEV(dev, &pdev->dev);
 	dev->netdev_ops = &rtl_netdev_ops;
 	tp = netdev_priv(dev);
