From 966471712184cfb3b067f2ae8dad9d8e2a896cae Mon Sep 17 00:00:00 2001
From: Bearice Ren <bearice@icybear.net>
Date: Tue, 20 Sep 2016 11:52:08 +0800
Subject: [PATCH] add filter-a & filter-aaaa option

---
 src/rfc1035.c | 20 ++++++++++
 1 files changed, 20 insertions(+)

--- a/src/rfc1035.c
+++ b/src/rfc1035.c
@@ -2111,6 +2111,26 @@ size_t answer_request(struct dns_header
 	    }
 	}
 
+      /* patch to filter a forwards */
+      if (qtype == T_A && option_bool(OPT_FILTER_A))
+        {
+          /* return a null reply */
+          ans = 1;
+          if (!dryrun)
+            log_query(F_CONFIG | F_IPV4 | F_NEG, name, &addr, NULL, 0);
+          break;
+        }
+
+      /* patch to filter aaaa forwards */
+      if (qtype == T_AAAA && option_bool(OPT_FILTER_AAAA))
+        {
+          /* return a null reply */
+          ans = 1;
+          if (!dryrun)
+            log_query(F_CONFIG | F_IPV6 | F_NEG, name, &addr, NULL, 0);
+          break;
+        }
+
       if (!ans)
 	return 0; /* failed to answer a question */
     }
