diff --git a/client/mysql.cc b/client/mysql.cc
index 822eeb1..4fafbdf 100644
--- a/client/mysql.cc
+++ b/client/mysql.cc
@@ -2684,7 +2684,7 @@ You can turn off this feature to get a quicker startup with -A\n\n");
         mysql_free_result(fields);
         break;
       }
-      field_names[i][num_fields*2]= '\0';
+      field_names[i][num_fields*2][0]= '\0';
       j=0;
       while ((sql_field=mysql_fetch_field(fields)))
       {
