--- a/net/Kconfig	2020-04-29 17:25:49.750444000 +0800
+++ b/net/Kconfig	2020-04-29 17:42:40.950424000 +0800
@@ -451,6 +451,18 @@
 	  migration of VMs with direct attached VFs by failing over to the
 	  paravirtual datapath when the VF is unplugged.
 
+config HW_NAT
+	bool "HW NAT support"
+	default n
+	---help---
+	 This feature provides a fast path to support network lan/wan nat.
+	 If you need hw_nat engine to reduce cpu loading, please say Y.
+
+	  Note that the answer to this question doesn't directly affect the
+	  kernel: saying N will just cause the configurator to skip all
+	  the questions about Mediatek Ethernet devices. If you say Y,
+	  you will be asked for your specific card in the following questions.
+
 endif   # if NET
 
 # Used by archs to tell that they support BPF JIT compiler plus which flavour.
--- a/net/Makefile	2020-04-23 16:36:46.000000000 +0800
+++ b/net/Makefile	2020-04-29 17:42:58.106487000 +0800
@@ -62,6 +62,9 @@
 obj-$(CONFIG_6LOWPAN)		+= 6lowpan/
 obj-$(CONFIG_IEEE802154)	+= ieee802154/
 obj-$(CONFIG_MAC802154)		+= mac802154/
+ifeq ($(CONFIG_HW_NAT),y)
+obj-y                           += nat/foe_hook/
+endif
 
 ifeq ($(CONFIG_NET),y)
 obj-$(CONFIG_SYSCTL)		+= sysctl_net.o
