diff --git a/net/dsa/tag_mtk.c b/net/dsa/tag_mtk.c
index c386fdc..e2bdd69 100644
--- a/net/dsa/tag_mtk.c
+++ b/net/dsa/tag_mtk.c
@@ -28,6 +28,8 @@ static struct sk_buff *mtk_tag_xmit(struct sk_buff *skb,
 	bool is_multicast_skb = is_multicast_ether_addr(dest) &&
 				!is_broadcast_ether_addr(dest);
 
+	skb_set_queue_mapping(skb, dp->index);
+
 	/* Build the special tag after the MAC Source Address. If VLAN header
 	 * is present, it's required that VLAN header and special tag is
 	 * being combined. Only in this way we can allow the switch can parse
