--- a/arch/arm64/boot/dts/qcom/ipq6018.dtsi
+++ b/arch/arm64/boot/dts/qcom/ipq6018.dtsi
@@ -189,34 +189,57 @@
 		#size-cells = <2>;
 		ranges;
 
-		rpm_msg_ram: memory@60000 {
+		rpm_msg_ram: rpm_msg_ram@60000 {
 			reg = <0x0 0x00060000 0x0 0x6000>;
 			no-map;
 		};
 
-		bootloader@4a100000 {
+		nss_region: nss@40000000 {
+			reg = <0x0 0x40000000 0x0 0x1000000>;
+			no-map;
+		};
+
+		bootloader_region: bootloader@4a100000 {
 			reg = <0x0 0x4a100000 0x0 0x400000>;
 			no-map;
 		};
 
-		sbl@4a500000 {
+		sbl_region: sbl@4a500000 {
 			reg = <0x0 0x4a500000 0x0 0x100000>;
 			no-map;
 		};
 
-		tz: memory@4a600000 {
+		tz_region: tz@4a600000 {
 			reg = <0x0 0x4a600000 0x0 0x400000>;
 			no-map;
 		};
 
-		smem_region: memory@4aa00000 {
+		smem_region: smem@4aa00000 {
 			reg = <0x0 0x4aa00000 0x0 0x100000>;
 			no-map;
 		};
 
-		q6_region: memory@4ab00000 {
-			reg = <0x0 0x4ab00000 0x0 0x5500000>;
+		q6_region: q6@4ab00000 {
+			reg = <0x0 0x4ab00000 0x0 0x4000000>;
+			no-map;
+		};
+
+		q6_etr_region: q6_etr@4eb00000 {
+			reg = <0x0 0x4eb00000 0x0 0x100000>;
+			no-map;
+		};
+
+		m3_dump_region: m3_dump@4ec00000 {
+			reg = <0x0 0x4ec00000 0x0 0x100000>;
+			no-map;
+		};
+
+		ramoops_region: ramoops@4ed00000 {
+			compatible = "ramoops";
+			reg = <0x0 0x4ed00000 0x0 0x100000>;
 			no-map;
+			record-size = <0x4000>;
+			console-size = <0x4000>;
 		};
 	};
 
