--- a/include/osdep_service_linux.h
+++ b/include/osdep_service_linux.h
@@ -220,18 +220,6 @@ typedef void *timer_hdl_context;
 	#define DMA_BIT_MASK(n) (((n) == 64) ? ~0ULL : ((1ULL<<(n))-1))
 #endif
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0))
-/* Porting from linux kernel commits 
-48eab831ae8b9f7002a533fa4235eed63ea1f1a3 
-3f6cffb8604b537e3d7ea040d7f4368689638eaf
-adeef3e32146a8d2a73c399dc6f5d76a449131b1
-*/
-static inline void eth_hw_addr_set(struct net_device *dev, const u8 *addr)
-{
-	memcpy(dev->dev_addr, addr, ETH_ALEN);
-}
-#endif
-
 typedef unsigned long systime;
 typedef struct tasklet_struct _tasklet;
 
