From Yukari <mistelinn@gmail.com>
Subject: Add tx_power check during eeprom loading
Date: Tue, 5 Feb 2025 16:38:46 +0800

Create tx_power_fix.patch
---

diff --git a/mt7996/eeprom.c b/mt7996/eeprom.c
index 53dfac02f..a699668ca 100644
--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -206,6 +206,13 @@ static int mt7996_eeprom_load(struct mt7996_dev *dev)
                                goto out;
                        }
                }
+
+               /* read tx_power values from fw */
+               u8 *eeprom = dev->mt76.eeprom.data;
+               if (!eeprom[MT_EE_TX0_POWER_2G] || !eeprom[MT_EE_TX0_POWER_5G] || !eeprom[MT_EE_TX0_POWER_6G] ) {
+                       use_default = true;
+                       goto out;
+               }
        }

 out:
