Signed-off-by: Kevin Darbyshire-Bryant <ldir at darbyshire-bryant.me.uk>
---
 service/instance.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/service/instance.c b/service/instance.c
index e872ba0..b78a65f 100644
--- a/service/instance.c
+++ b/service/instance.c
@@ -800,7 +800,8 @@
 
 	r = stat(UJAIL_BIN_PATH, &s);
 	if (r < 0) {
-		ERROR("unable to find %s: %m (%d)\n", UJAIL_BIN_PATH, r);
+		ULOG_WARN("Cannot jail service %s::%s. %s: %m (%d) Are jails enabled?\n",
+				in->srv->name, in->name, UJAIL_BIN_PATH, r);
 		return 0;
 	}
 
