--- a/src/i40e_virtchnl_pf.c
+++ b/src/i40e_virtchnl_pf.c
@@ -6661,7 +6661,6 @@ int i40e_get_vf_stats(struct net_device
 #endif /* HAVE_VF_STATS */
 #endif /* IFLA_VF_MAX */
 #ifdef HAVE_NDO_SET_VF_LINK_STATE
-#ifdef CONFIG_DCB
 #ifdef CONFIG_PCI_IOV
 /**
  * i40e_configure_vf_link
@@ -6739,6 +6738,7 @@ error_out:
 	return ret;
 }
 
+#ifdef CONFIG_DCB
 /**
  * i40e_enable_vf_queues
  * @vsi: PCI device information struct
@@ -6792,8 +6792,8 @@ int i40e_enable_vf_queues(struct i40e_vs
 err_out:
 	return ret;
 }
-#endif /* PCI_IOV */
 #endif /* CONFIG_DCB */
+#endif /* PCI_IOV */
 /**
  * i40e_get_vlan_anti_spoof
  * @pdev: PCI device information struct
