--- a/arch/arm64/boot/dts/qcom/ipq8074.dtsi
+++ b/arch/arm64/boot/dts/qcom/ipq8074.dtsi
@@ -86,17 +86,27 @@
 		#size-cells = <2>;
 		ranges;
 
-		bootloader@4a600000 {
+		nss_region: nss@40000000 {
+			reg = <0x0 0x40000000 0x0 0x1000000>;
+			no-map;
+		};
+
+		tzapp_region: tzapp@4a400000 {
+			reg = <0x0 0x4a400000 0x0 0x200000>;
+			no-map;
+		};
+
+		bootloader_region: bootloader@4a600000 {
 			reg = <0x0 0x4a600000 0x0 0x400000>;
 			no-map;
 		};
 
-		sbl@4aa00000 {
+		sbl_region: sbl@4aa00000 {
 			reg = <0x0 0x4aa00000 0x0 0x100000>;
 			no-map;
 		};
 
-		smem@4ab00000 {
+		smem_region: smem@4ab00000 {
 			compatible = "qcom,smem";
 			reg = <0x0 0x4ab00000 0x0 0x100000>;
 			no-map;
@@ -104,10 +114,33 @@
 			hwlocks = <&tcsr_mutex 3>;
 		};
 
-		memory@4ac00000 {
+		memory_region: memory@4ac00000 {
 			reg = <0x0 0x4ac00000 0x0 0x400000>;
 			no-map;
 		};
+
+		q6_region: q6@4b000000 {
+			reg = <0x0 0x4b000000 0x0 0x4000000>;
+			no-map;
+		};
+
+		q6_etr_region: q6_etr@4f000000 {
+			reg = <0x0 0x4f000000 0x0 0x100000>;
+			no-map;
+		};
+
+		m3_dump_region: m3_dump@4f100000 {
+			reg = <0x0 0x4f100000 0x0 0x100000>;
+			no-map;
+		};
+
+		ramoops_region: ramoops@4f200000 {
+			compatible = "ramoops";
+			reg = <0x0 0x4f200000 0x0 0x100000>;
+			no-map;
+			record-size = <0x4000>;
+			console-size = <0x4000>;
+		};
 	};
 
 	firmware {
