--- a/Makefile
+++ b/Makefile
@@ -1,11 +1,11 @@
-OS = windows
+OS = linux
 BUILD_TYPE = release_static
 
 OUTPUT := ps3netsrv
 OBJS = src/main.o src/padlock.o src/aes.o src/compat.o src/File.o src/VIsoFile.o
 
-CFLAGS = -Wall -Wno-format -I./include -std=gnu99 -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DPOLARSSL
-CPPFLAGS += -Wall -Wno-format -I./include -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DPOLARSSL
+CFLAGS = -Wall -I./include -std=gnu99 -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DPOLARSSL
+CPPFLAGS += -Wall -I./include -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DPOLARSSL
 
 #CFLAGS += -Doff64_t=off_t
 #CPPFLAGS += -Doff64_t=off_t
@@ -18,7 +18,7 @@ LDFLAGS = -L.
 LIBS = -lstdc++
 
 ifeq ($(OS), linux)
-LIBS += -lpthread
+LIBS += -lpthread -lgcc_eh
 endif
 
 ifeq ($(OS), windows)
