--- a/Makefile
+++ b/Makefile
@@ -416,7 +416,7 @@ PYTHON3		?= python3
 
 # The devicetree compiler and pylibfdt are automatically built unless DTC is
 # provided. If DTC is provided, it is assumed the pylibfdt is available too.
-DTC_INTREE	:= $(objtree)/scripts/dtc/dtc
+DTC	:= $(objtree)/scripts/dtc/dtc
 DTC		?= $(DTC_INTREE)
 DTC_MIN_VERSION	:= 010406
 
@@ -2032,7 +2032,7 @@ endif
 # Check dtc and pylibfdt, if DTC is provided, else build them
 PHONY += scripts_dtc
 scripts_dtc: scripts_basic
-	$(Q)if test "$(DTC)" = "$(DTC_INTREE)"; then \
+	$(Q)if test "$(DTC)" = "$(DTC)"; then \
 		$(MAKE) $(build)=scripts/dtc; \
 	else \
 		if ! $(DTC) -v >/dev/null; then \
