--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -148,6 +148,7 @@ dtb-$(CONFIG_ROCKCHIP_RK3399) += \
 	rk3399-firefly.dtb \
 	rk3399-gru-bob.dtb \
 	rk3399-gru-kevin.dtb \
+	rk3399-guangmiao-g4c.dtb \
 	rk3399-khadas-edge.dtb \
 	rk3399-khadas-edge-captain.dtb \
 	rk3399-khadas-edge-v.dtb \
@@ -158,6 +159,7 @@ dtb-$(CONFIG_ROCKCHIP_RK3399) += \
 	rk3399-nanopi-m4b.dtb \
 	rk3399-nanopi-neo4.dtb \
 	rk3399-nanopi-r4s.dtb \
+	rk3399-nanopi-r4se.dtb \
 	rk3399-orangepi.dtb \
 	rk3399-pinebook-pro.dtb \
 	rk3399-pinephone-pro.dtb \
@@ -182,12 +184,17 @@ dtb-$(CONFIG_ROCKCHIP_RK3568) += \
 	rk3566-soquartz-model-a.dtb \
 	rk3568-bpi-r2-pro.dtb \
 	rk3568-evb.dtb \
+	rk3568-fastrhino-r66s.dtb \
+	rk3568-fastrhino-r68s.dtb \
 	rk3568-generic.dtb \
 	rk3568-lubancat-2.dtb \
+	rk3568-mrkaio-m68s.dtb \
 	rk3568-nanopi-r5c.dtb \
 	rk3568-nanopi-r5s.dtb \
 	rk3568-odroid-m1.dtb \
+	rk3568-photonicat.dtb \
 	rk3568-radxa-e25.dtb \
+	rk3568-roc-pc.dtb \
 	rk3568-rock-3a.dtb
 
 dtb-$(CONFIG_ROCKCHIP_RK3588) += \
@@ -200,7 +207,9 @@ dtb-$(CONFIG_ROCKCHIP_RK3588) += \
 	rk3588-quartzpro64.dtb \
 	rk3588-turing-rk1.dtb \
 	rk3588s-rock-5a.dtb \
-	rk3588-rock-5b.dtb
+	rk3588-rock-5b.dtb \
+	rk3588s-nanopi-r6c.dtb \
+	rk3588s-nanopi-r6s.dtb
 
 dtb-$(CONFIG_ROCKCHIP_RV1108) += \
 	rv1108-elgin-r1.dtb \
