--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -96,6 +96,19 @@ dtb-$(CONFIG_ROCKCHIP_RK3368) += \
 	rk3368-geekbox.dtb \
 	rk3368-px5-evb.dtb \
 
+dtb-$(CONFIG_ROCKCHIP_RK3399) += \
+	rk3399-guangmiao-g4c.dtb \
+	rk3399-nanopi-r4se.dtb
+
+dtb-$(CONFIG_ROCKCHIP_RK3568) += \
+	rk3568-mrkaio-m68s.dtb \
+	rk3568-photonicat.dtb
+
+dtb-$(CONFIG_ROCKCHIP_RK3588) += \
+	rk3588-armsom-sige7.dtb \
+	rk3588s-nanopi-r6c.dtb \
+	rk3588s-nanopi-r6s.dtb
+
 dtb-$(CONFIG_ARCH_S5P4418) += \
 	s5p4418-nanopi2.dtb
 
