diff --git a/src/main.c b/src/main.c
index 6a9c97a..d902b3d 100644
--- a/src/main.c
+++ b/src/main.c
@@ -630,7 +630,7 @@ ndpi_mt(const struct sk_buff *skb, struct xt_action_param *par)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3,0,0)
 	} else if (nf_ct_is_untracked(skb)){
 #else
-	} else if (nf_ct_is_untracked(ct)){
+	} else if (nf_ct_is_expired(ct)){
 #endif
 		pr_info ("xt_ndpi: ignoring untracked sk_buff.\n");
 		return false;               
