--- a/mt_wifi_ap/Makefile
+++ b/mt_wifi_ap/Makefile
@@ -12,8 +12,8 @@ EXTRA_CFLAGS = -I$(WIFI_DRV_PATH)/mt_wifi/include \
 				-I$(WIFI_DRV_PATH)/mt_wifi/ate/include \
 				-I$(WIFI_DRV_PATH)/mt_wifi/include/mcu \
 				-I$(WIFI_DRV_PATH)/mt_wifi/embedded/dbg_txcmd/include \
-				-I$(WIFI_DRV_PATH)/warp_driver/warp \
-				-I$(WIFI_DRV_PATH)/warp_driver/warp/mcu
+				-I$(WIFI_DRV_PATH)/../warp \
+				-I$(WIFI_DRV_PATH)/../warp/mcu
 
 ifeq ($(CONFIG_SUPPORT_OPENWRT),y)
 GLOBAL_DEFINES = include/uapi/linux/wapp
-- 
2.25.1

